envisage/plugins/ipython_kernel/internal_ipkernel.py

```diff
@@ -3,17 +3,43 @@
 https://github.com/ipython/ipython/blob/2.x/examples/Embedding/internal_ipkernel.py
 
 """
-from distutils.version import StrictVersion as Version
-import sys
+from __future__ import absolute_import, print_function, unicode_literals
 
-import ipykernel
-from ipykernel.connect import connect_qtconsole
-from ipykernel.kernelapp import IPKernelApp
-from tornado import ioloop
+import logging
 
+import ipykernel.connect
+import IPython.utils.io
+import six
+
+from envisage.plugins.ipython_kernel.kernelapp import IPKernelApp
 from traits.api import Any, HasStrictTraits, Instance, List
 
-NEEDS_IOLOOP_PATCH = Version(ipykernel.__version__) >= Version('4.7.0')
+logger = logging.getLogger(__name__)
+
+
+# Replacements for IPython.utils.io.raw_print and IPython.utils.raw_print_err
+# that send output to logging instead of to __stdout__ and __stderr__.
+
+def log_print(*args, **kw):
+    """
+    Logging replacement for IPython.utils.io.raw_print.
+    Instead of printing to __stdout__, this function logs at INFO level.
+    """
+    message = six.StringIO()
+    print(*args, sep=kw.get('sep', ' '), end=kw.get('end', '\n'),
+          file=message)
+    logger.info(message.getvalue())
+
+
+def log_print_err(*args, **kw):
+    """
+    Logging replacement for IPython.utils.io.raw_print_err.
+    Instead of printing to __stdout__, this function logs at WARNING level.
+    """
+    message = six.StringIO()
+    print(*args, sep=kw.get('sep', ' '), end=kw.get('end', '\n'),
+          file=message)
+    logger.warning(message.getvalue())
```

Comment:
`warning` or `error`?

Also this method and `log_print` are awfully similar.

Reply:
Warning, I think, but I'm not 100% convinced either way. The implicit message here is "the IPython machinery spewed some stuff to stderr; here it is. It *might* be harmless, but it's probably worth investigating." That says "warning" rather than "error" to me.

It's probably academic anyway: I can't find a single use of `raw_print_err` (or its `rprinte` alias) within any of the IPython packages we're currently using.

Yes, they're similar, but they're so short that I think factoring out the repetition would make this harder to read rather than easier.

Reply:
I'm also considering using a custom logger name to make it clearer where the output is coming from in the logs. Something like `"ipython.stream_output"` instead of the current `"envisage.plugins.ipython_kernel.internal_ipkernel`, which is a bit obscure.

Reply:
One more thing: it may be worth adding a couple of direct tests for these two functions with bytes and Unicode arguments (and with a mixture of the two). It's all fine and clear cut on Python 3; less so on Python 2. If you replace `six.StringIO` with `io.StringIO`, for example, things will fail on Python 2, so there should be a test that also fails in that case.